contracts/UnionZap.sol

```diff
@@ -319,64 +361,211 @@ contract UnionZap is Ownable, UnionBase {
+        } else {
+            (bool success, ) = tokenInfo[_token].swapper.call{value: _amount}(
+                ""
+            );
+            require(success, "ETH transfer failed");
+            ISwapper(tokenInfo[_token].swapper).buy(_amount);
+        }
+    }
+
+    /// @notice Swap or lock all CRV for cvxCRV
+    /// @param _minAmountOut - the min amount of cvxCRV expected
+    /// @param _lock - whether to lock or swap
+    /// @return the amount of cvxCrv obtained
+    function _toCvxCrv(uint256 _minAmountOut, bool _lock)
+        internal
+        returns (uint256)
+    {
+        uint256 _crvBalance = IERC20(CRV_TOKEN).balanceOf(address(this));
         // swap on Curve if there is a premium for doing so
-        if (!lock) {
-            _swapCrvToCvxCrv(_crvBalance, address(this), minAmountOut);
+        if (!_lock) {
+            return _swapCrvToCvxCrv(_crvBalance, address(this), _minAmountOut);
         }
         // otherwise deposit & lock
-        else {
-            ICvxCrvDeposit(CVXCRV_DEPOSIT).deposit(_crvBalance, true);
-            _locked = true;
-        }
-
-        uint256 _cvxCrvBalance = IERC20(CVXCRV_TOKEN).balanceOf(address(this));
-
-        // freeze distributor before transferring funds
-        IMerkleDistributorV2(unionDistributor).freeze();
-
-        // estimate gas cost
-        uint256 _gasUsed = _startGas -
-            gasleft() +
-            BASE_TX_GAS +
-            16 *
-            msg.data.length +
-            FINAL_TRANSFER_GAS;
-        // compute the ETH/CRV exchange rate based on previous curve swap
-        uint256 _gasCostInCrv = (_gasUsed * tx.gasprice * _swappedCrv) /
-            _ethBalance;
+        // slippage check
+        assert(_crvBalance > _minAmountOut);
+        ICvxCrvDeposit(CVXCRV_DEPOSIT).deposit(_crvBalance, true);
+        return _crvBalance;
+    }
 
-        uint256 _netDeposit = _cvxCrvBalance - _gasCostInCrv;
+    /// @notice Splits contract balance into output tokens as per weights
+    /// @param lock - whether to lock or swap crv to cvxcrv
+    /// @param weights - weight of output assets (cvxCRV, FXS, CVX) in bips
+    /// @param minAmounts - min amount out of each output token (cvxCRV for CRV)
+    /// @dev weights must sum to 10000
+    function adjust(
+        bool lock,
+        uint16[] calldata weights,
+        uint256[] calldata minAmounts
+    ) public onlyOwner validWeights(weights) {
+        require(
+            minAmounts.length == outputTokens.length,
+            "Invalid min amounts"
+        );
+        // start calculating the allocations of output tokens
+        uint256 _totalEthBalance = address(this).balance;
+
+        uint256[] memory prices = new uint256[](outputTokens.length);
+        uint256[] memory amounts = new uint256[](outputTokens.length);
+        address _outputToken;
+
+        // first loop to calculate total ETH amounts and store oracle prices
+        for (uint256 i; i < weights.length; ++i) {
```

Comment:
Only methods that use the `validWeights` modifier and have 1+ loops w/ loop conditions based on `weights.length`: it should be more gas efficient to have the weight-checking logic in the loop (the modifier would be removed and there'd be 1 less set of iterations over `weights`)